src/nginx/grpc.cc

```diff
@@ -94,20 +97,85 @@ std::pair<Status, std::shared_ptr<::grpc::GenericStub>> GrpcGetStub(
     return std::make_pair(Status::OK, it->second);
   }
 
+  ngx_str_t ngx_addr = ngx_std_to_str_unsafe(address);
+
   ::grpc::ChannelArguments channel_arguments;
 
   channel_arguments.SetMaxReceiveMessageSize(INT_MAX);
   channel_arguments.SetMaxSendMessageSize(INT_MAX);
 
-  auto result =
-      std::make_shared<::grpc::GenericStub>(::grpc::CreateCustomChannel(
-          address, ::grpc::InsecureChannelCredentials(), channel_arguments));
+  if (ngx_string_startswith(ngx_addr, kGrpcTlsPrefix)) {
```

Comment:
if the `address` begins with `grpcs://`... then turn on upstream TLS, optionally with `grpc_roots` set (or not, in which case defaults are used). if a client key and cert are set, pass those in, too.